output/cloud/expv2/output.go

```diff
@@ -151,42 +171,52 @@ func (o *Output) flushMetrics() {
 
 	err := o.metricsFlusher.Flush(ctx)
 	if err != nil {
-		o.logger.WithError(err).Error("Failed to push metrics to the cloud")
-
-		if o.shouldStopSendingMetrics(err) {
-			o.logger.WithError(err).Warn("Interrupt sending metrics to cloud due to an error")
-			serr := errext.WithAbortReasonIfNone(
-				errext.WithExitCodeIfNone(err, exitcodes.ExternalAbort),
-				errext.AbortedByOutput,
-			)
-			if o.config.StopOnError.Bool {
-				o.testStopFunc(serr)
-			}
-			close(o.stopMetricsCollection)
-		}
+		o.handleFlushError(err)
 		return
 	}
 
 	o.logger.WithField("t", time.Since(start)).Debug("Successfully flushed buffered samples to the cloud")
 }
 
-// shouldStopSendingMetrics returns true if the output should interrupt the metric flush.
+// handleFlushError handles errors generated from the flushing operation.
+// It may interrupts the metric collection or invokes abort of the test.
 //
-// note: The actual test execution should continues,
-// since for local k6 run tests the end-of-test summary (or any other outputs) will still work,
+// note: The actual test execution should continue, since for local k6 run tests
+// the end-of-test summary (or any other outputs) will still work,
 // but the cloud output doesn't send any more metrics.
-// Instead, if cloudapi.Config.StopOnError is enabled
-// the cloud output should stop the whole test run too.
-// This logic should be handled by the caller.
-func (o *Output) shouldStopSendingMetrics(err error) bool {
-	if err == nil {
-		return false
+// Instead, if cloudapi.Config.StopOnError is enabled the cloud output should
+// stop the whole test run too. This logic should be handled by the caller.
+func (o *Output) handleFlushError(err error) {
+	o.logger.WithError(err).Error("Failed to push metrics to the cloud")
+
+	var errResp cloudapi.ErrorResponse
+	if !errors.As(err, &errResp) || errResp.Response == nil {
+		return
 	}
-	if errResp, ok := err.(cloudapi.ErrorResponse); ok && errResp.Response != nil { //nolint:errorlint
-		// The Cloud service returns the error code 4 when it doesn't accept any more metrics.
-		// So, when k6 sees that, the cloud output just stops prematurely.
-		return errResp.Response.StatusCode == http.StatusForbidden && errResp.Code == 4
+
+	// The Cloud service returns the error code 4 when it doesn't accept any more metrics.
+	// So, when k6 sees that, the cloud output just stops prematurely.
+	if errResp.Response.StatusCode != http.StatusForbidden || errResp.Code != 4 {
+		return
 	}
 
-	return false
+	o.logger.WithError(err).Warn("Interrupt sending metrics to cloud due to an error")
+
+	// Do not close multiple times (that would panic) in the case
+	// we hit this multiple times and/or concurrently
+	select {
+	case <-o.stopSamplesCollection:
+		return
+	default:
+		close(o.stopSamplesCollection)
+	}
```

Comment:
While this will work as `handleFlushError` is only called in flush and that is called only in one placea at a time.

I kind of feel we are doing this in the wrong place - if there was an error that will abort teh flushing of metrics we should stop *flushing* metrics. Not only collecting them.

Reply:
Do we want to check this all the time when the expectation for this is to get it very few times? Why should we continuously sync over this select statement for a successful test where we will never hit it?

Reply:
Why should we make a network call trhat will fail ?

Reply:
Also in practice you can just have it in the `Stop` method I guess - no need to go through all the different parts if we are not going to flush anything.

Reply:
>Also in practice you can just have it in the Stop method I guess 

The problem with this is when we will move to have multiple-concurrent flushers then we could get this function called multiple times.

Reply:
I think we may try to have this handler as a goroutine and get the error on a channel in the classic go way. In this way, we have a centralized place for doing the close.

Reply:
We do something like this in [the v1](https://github.com/grafana/k6/blob/master/output/cloud/v1/output.go) maybe more of the structure should be copied from there

Reply:
I tried but I don't think it is cleaner. How does it achieve the goal?! Invoking the `flush` operation from the aggregation goroutine that it sounds not optimal.

I'm not against doing the refactor and I am happy to improve it, but it doesn't sound to me like a priority right now. Performances should be fine and the logic has been fixed not panicking. I included it in the reminders list.

I would like to merge the chain before and the retry + concurrent flushers then with a better picture we can do the refactor.

Also, the metadata flush will be added soon, so the risk is to refactor this multiple times in a few days.